### PR TITLE
Reorganize reporting and enable error reporting

### PR DESCRIPTION
**Pull request by murrant**
Move code to ErrorReportingProvider
 Enable reporting of error (and warning) messages.
 report module exceptions
set flare key in the normal place.  I confirmed early (before Flare::filterExceptionsUsing is called) and late (after Flare is booted) both behave correctly.

### CODE REVIEW

File: app/Exceptions/Handler.php [hunk 1]
```diff
@@ -63,6 +63,12 @@ public function render($request, Throwable $exception)
         return parent::render($request, $exception);
     }
 
+    public function report(Throwable $e)
+    {
+        parent::report($e);
+        \Flare::report($e);
```
Comment (Jellyfrog) [inline app/Exceptions/Handler.php:69]: Is this really needed?

Comment (murrant) [reply app/Exceptions/Handler.php:69]: I think so, easy to test.  If I remember, when calling report() it was not sending it.

Comment (murrant) [reply app/Exceptions/Handler.php]: Not needed, removed :)

File: app/Providers/ErrorReportingProvider.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+class ErrorReportingProvider extends \Facade\Ignition\IgnitionServiceProvider
+{
+    protected $errorReportingLevel = E_ALL & ~E_NOTICE;
+    private $laravelErrorHandler;
+
+    public function boot()
+    {
+        // don't report when:
+        Flare::filterExceptionsUsing(function (\Exception $e) {
```
Comment (Jellyfrog) [inline app/Providers/ErrorReportingProvider.php:44]: You switched from `filterReportsUsing` to `filterExceptionsUsing`, which indeed happens earlier, but it won't catch log-message logging (if we would enable that).

Maybe we should convert these checks to a static function, and set both  `filterReportsUsing` and `filterExceptionsUsing`, for good measure :)

File: app/Providers/ErrorReportingProvider.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+use ErrorException;
+use Facade\FlareClient\Report;
+use Facade\Ignition\Facades\Flare;
+use Illuminate\Support\Str;
+use LibreNMS\Config;
+use LibreNMS\Util\Git;
+use LibreNMS\Util\Version;
+
+class ErrorReportingProvider extends \Facade\Ignition\IgnitionServiceProvider
```
Comment (Jellyfrog) [inline app/Providers/ErrorReportingProvider.php:36]: Clever :)

File: app/Providers/ErrorReportingProvider.php (hunk 2 above)
Comment (Jellyfrog) [reply app/Providers/ErrorReportingProvider.php:44]: There is a bug with filterReportsUsing , the flare client will think it sent a report and generate an UUID, will fix that upstream some day.. so using filterExceptionsUsing is good also :P

File: app/Providers/ErrorReportingProvider.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+                ]);
+            } catch (\Exception $e) {
+            }
+
+            return $next($report);
+        });
+
+        // Override the Laravel error handler
+        $this->laravelErrorHandler = set_error_handler([$this, 'handleError']);
```
Comment (Jellyfrog) [inline app/Providers/ErrorReportingProvider.php:127]: > Returns the previously defined error handler (if any).

Oh, this is weird syntax.

File: app/Providers/ErrorReportingProvider.php (hunk 2 above)
Comment (murrant) [reply app/Providers/ErrorReportingProvider.php:44]: Setting both seems fine to me.  From what I could sus of the flow filterExceptionsUsing is always called currently.

File: app/Providers/ErrorReportingProvider.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+     * @return void
+     *
+     * @throws \ErrorException
+     */
+    public function handleError($level, $message, $file = '', $line = 0, $context = [])
+    {
+        // report errors if they are allowed
+        if ($this->errorReportingLevel & $level) {
+            Flare::report(new ErrorException($message, 0, $level, $file, $line));
+        }
```
[the remark this answers is not in the document]
Comment (murrant) [reply app/Providers/ErrorReportingProvider.php:149]: It cannot.  It must first be converted to an ErrorException by the default Laravel hander.  We deleted the default handler, so we must do the work.

File: app/Providers/ErrorReportingProvider.php (hunk 4 above)
Comment (murrant) [reply app/Providers/ErrorReportingProvider.php:127]: It is but, it did prevent me from having to duplicate the default Laravel handler code or do some VERY strange stuff.

File: app/Providers/ErrorReportingProvider.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (murrant) [reply app/Providers/ErrorReportingProvider.php:149]: Hmm, but we could change it to always call flare report with a new ErrorException.

That does not necessarily seem better to me though.

File: app/Providers/ErrorReportingProvider.php (hunk 2 above)
Comment (Jellyfrog) [reply app/Providers/ErrorReportingProvider.php:44]: It's not called if it's a log message (that feature is disabled tho)

**Pull request merged**